test/unit_test/app_common/test_storage.py

```diff
@@ -45,50 +46,58 @@ def random_meta():
 
 
-tmp_dir = tempfile.TemporaryDirectory()
-tmp_dir_name = tmp_dir.name
+tmp_dir = None
+storages = defaultdict()
+minio_ps = None
 
 
-@classmethod
-def teardown_class(cls):
-    tmp_dir.cleanup()
+@pytest.mark.xdist_group(name="storage_tests_group")
+@pytest.mark.parametrize("storage", [FilesystemStorage, S3Storage])
+class TestStorage:
 
+    @classmethod
+    def setup_class(cls):
+        global tmp_dir
+        global minio_ps
 
-@pytest.mark.parametrize(
-    "storage",
-    [
-        FilesystemStorage(root_dir=tmp_dir_name),
-        # S3Storage(endpoint="{client_ip_addr}:{api_port}", access_key="admin", secret_key="password", secure=False, bucket_name="nvflare-storage"),
-    ],
-)
-class TestStorage:
-    """
-    To test filesystem implementation:
-    ex: FilesystemStorage(root_dir="/tmp/storage_tests/filesystem_test")
-
-    To test NFS implementation:
-    ex: FilesystemStorage(root_dir="/tmp/storage_tests/nfs_test")
-        set up NFS server:
-            sudo apt install nfs-kernel-server
-            add in /etc/exports: /home/{user}/nfs_server {client_ip_addr}(rw,sync,no_subtree_check)
-            sudo exportfs -a
-            sudo systemctl restart nfs-kernel-server
-        setup NFS client:
-            sudo apt install nfs-common
-            mount -t nfs -o nfsvers=3 {server_ip_addr}:/home/{user}/nfs_server /tmp/storage_tests/nfs_test
-
-    To test S3 implementation using Minio server:
-    ex: S3Storage(endpoint="{client_ip_addr}:{api_port}", access_key="admin", secret_key="password", secure=False, bucket_name="nvflare-storage")
-      wget https://dl.min.io/server/minio/release/linux-amd64/minio
-      chmod +x minio
-      sudo MINIO_ROOT_USER=admin MINIO_ROOT_PASSWORD=password ./minio server /home/seyang/data/storage_tests/s3_test --console-address ":{console_port}"
-
-    """
-
-    @pytest.mark.parametrize("n_files", [20, 200])
+        tmp_dir = tempfile.TemporaryDirectory()
+        tmp_dir_name = tmp_dir.name
+
+        # start local minio server as a compatible S3 bucket for testing
+
+        commands = [
+            ["wget", "https://dl.min.io/server/minio/release/linux-amd64/minio"],
+            ["chmod", "+x", "minio"],
+            ["./minio", "server", os.path.join(tmp_dir_name, "s3-storage"), "--console-address", ":9001"]
+        ]
+
+        env = os.environ.update({"MINIO_ROOT_USER": "admin", "MINIO_ROOT_PASSWORD": "password"})
+        for command in commands:
+            minio_ps = subprocess.Popen(command, stderr=subprocess.PIPE, stdout=subprocess.PIPE, env=env)
+            if command[0] != "./minio":
+                minio_ps.communicate()
+
+        time.sleep(15)
+
+        storages[FilesystemStorage] = FilesystemStorage(root_dir=os.path.join(tmp_dir_name, "filesystem-storage"))
+        storages[S3Storage] = S3Storage(endpoint="localhost:9000", access_key="admin", secret_key="password", secure=False, bucket_name="nvflare-storage")
```

Comment:
suggest using "string" as key instead of class
so `storages["FilesystemStorage"]`